File: frontend/src/api/hooks/system.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { api } from '@/api/client';
import { qk } from '@/api/keys';
import { useAuthStore } from '@/stores/auth';
import type {
  AuditEntry,
  AuditQuery,
  ConnectionTestRequest,
  ConnectionTestResponse,
  GeneratedClusterConfig,
  InfoResponse,
  LoginRequest,
  LoginResponse,
  Page,
  User,
} from '@/api/types';

export function useInfo() {
  return useQuery({
    queryKey: qk.info(),
    queryFn: () => api.get<InfoResponse>('/info'),
    staleTime: 60_000,
    retry: 1,
  });
}

export function useMe(enabled = true) {
  return useQuery({
    queryKey: qk.me(),
    queryFn: () => api.get<User>('/auth/me'),
    enabled,
    retry: false,
    staleTime: 60_000,
  });
}

export function useLogin() {
  const setSession = useAuthStore((s) => s.setSession);
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: LoginRequest) => api.post<LoginResponse>('/auth/login', body),
    onSuccess: (data) => {
      setSession(data.token, data.user);
      void qc.invalidateQueries();
    },
  });
}

export function useLogout() {
  const clear = useAuthStore((s) => s.clear);
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async () => {
      try {
        await api.post<void>('/auth/logout');
      } catch {
        /* logging out locally is enough */
      }
    },
    onSettled: () => {
      clear();
      qc.clear();
    },
  });
}

/**
 * Probe candidate cluster connection details (admin only). Changes nothing on the
 * server: the cluster inventory is deployment-managed, so the response also carries the
 * YAML fragment to apply. Not cached — a test is an explicit user action.
 */
export function useConnectionTest() {
  return useMutation({
    mutationFn: (body: ConnectionTestRequest) =>
      api.post<ConnectionTestResponse>('/system/connection-test', body),
  });
}

/** Render the k-shui.yaml fragment for these details without probing anything. */
export function useConnectionConfig() {
  return useMutation({
    mutationFn: (body: ConnectionTestRequest) =>
      api.post<GeneratedClusterConfig>('/system/connection-config', body),
  });
}

export function useAudit(query: AuditQuery) {
  return useQuery({
    queryKey: qk.audit(query),
    queryFn: () =>
      api.get<Page<AuditEntry>>('/audit', {
        page: query.page,
        perPage: query.perPage,
        clusterId: query.clusterId,
        user: query.user,
        action: query.action,
      }),
    placeholderData: (prev) => prev,
  });
}

File: backend/k_shui/api/routers/system.py
"""System information."""

from __future__ import annotations

import time

from fastapi import APIRouter, Depends, Request

from k_shui import __version__
from k_shui.api.schemas.connection import (
    ConnectionTestRequest,
    ConnectionTestResponse,
    GeneratedConfig,
)
from k_shui.api.schemas.security import SystemInfo
from k_shui.config import Settings
from k_shui.core.audit import audit
from k_shui.core.auth import Principal, non_mutating, optional_principal, require_admin
from k_shui.core.connection_test import generate_config, run_connection_test
from k_shui.core.registry import ClusterRegistry, get_registry, get_settings

router = APIRouter(tags=["system"])


@router.get("/info", response_model=SystemInfo)
async def info(
    request: Request,
    settings: Settings = Depends(get_settings),
    registry: ClusterRegistry = Depends(get_registry),
    principal: Principal | None = Depends(optional_principal),
) -> SystemInfo:
    """Bootstrap document for the UI.

    This is the one endpoint an unauthenticated client may read: it has to learn *how*
    to sign in before it can. Anonymous callers therefore get the auth block and nothing
    else — no cluster inventory, no feature flags.
    """
    started = getattr(request.app.state, "started_at", time.time())
    if principal is None:
        return SystemInfo(
            version=__version__,
            uptimeSeconds=round(time.time() - started, 3),
            auth={"type": settings.auth.type, "enabled": True, "user": None},
            features={},
            clusters=[],
            readOnly=settings.server.readOnly,
            basePath=settings.server.basePath,
        )
    features: dict[str, bool] = {}
    for ctx in registry.all():
        for key, enabled in ctx.config.features.items():
            features[key] = features.get(key, False) or enabled
    features["alerts"] = True
    features["audit"] = True
    features["metrics"] = settings.telemetry.metrics
    return SystemInfo(
        version=__version__,
        uptimeSeconds=round(time.time() - started, 3),
        auth={
            "type": settings.auth.type,
            "enabled": settings.auth.type != "none",
            "user": principal.to_dict(),
        },
        features=features,
        clusters=[{"id": c.config.id, "name": c.config.name or c.config.id} for c in registry.all()],
        readOnly=settings.server.readOnly,
        basePath=settings.server.basePath,
    )


@router.post("/system/connection-test", response_model=ConnectionTestResponse)
@non_mutating
async def connection_test(
    body: ConnectionTestRequest,
    request: Request,
    principal: Principal = Depends(require_admin),
) -> ConnectionTestResponse:
    """Probe candidate connection details for a cluster that is not configured yet.

    Every *provided* component is tested independently with a bounded timeout, and each
    result says whether the problem is connectivity, credentials, permissions or
    configuration. Nothing is written: the cluster inventory is deployment-managed, so
    the response also carries the YAML fragment to apply (credentials as ``${ENV_VAR}``).
    Submitted secrets are never echoed back.
    """
    result = await run_connection_test(body)
    await audit(
        request,
        "system.connection_test",
        resource=body.clusterId,
        details={
            "bootstrapServers": body.bootstrapServers,
            "ok": result.ok,
            "results": {c.component: c.status for c in result.components},
        },
        cluster_id=body.clusterId,
    )
    return result


@router.post("/system/connection-config", response_model=GeneratedConfig)
@non_mutating
async def connection_config(
    body: ConnectionTestRequest,
    principal: Principal = Depends(require_admin),
) -> GeneratedConfig:
    """Render the ``k-shui.yaml`` cluster fragment for these details without probing anything."""
    return generate_config(body)
